src/plonkish/ir/pil.rs

```diff
@@ -0,0 +1,180 @@
+use crate::{
+    ast::{query::Queriable, Circuit},
+    poly::Expr,
+    util::UUID,
+    wit_gen::TraceWitness,
+};
+use std::{collections::HashMap, fmt::Debug};
+extern crate regex;
+
+// PIL circuit IR
+pub struct PilCircuit<F> {
+    // circuit level
+    pub circuit_name: String,
+    pub num_steps: usize,
+
+    // circuit level - col witness
+    pub col_witness: Vec<UUID>, // internal signal NOT dedupped
+
+    // circuit level - col fixed
+    pub fixed_signals: HashMap<UUID, Vec<F>>, // fixed signal UUID -> fixed assignments vector
+    pub step_types_instantiations: HashMap<UUID, Vec<usize>>, /* step type UUID -> step
+                                               * instances
+                                               * vector {0, 1}^num_steps */
+    pub first_step: Option<UUID>,
+    pub last_step: Option<UUID>,
+
+    // step type level
+    pub step_types: Vec<UUID>,
+    pub constraints: HashMap<UUID, Vec<Expr<F, Queriable<F>>>>, // step type UUID -> constraints
+    pub transitions: HashMap<UUID, Vec<Expr<F, Queriable<F>>>>,
+    pub lookups: HashMap<UUID, Vec<PilLookup<F>>>, // step type UUID -> lookups vector
+}
+
+type PilLookup<F> = Vec<(Expr<F, Queriable<F>>, Expr<F, Queriable<F>>)>;
+
+impl<F: Clone + Debug> PilCircuit<F> {
+    pub fn from_ast_and_witness<TraceArgs>(
```

Comment:
@qwang98 this function is too long, please slit it.

Reply:
Converting AST to IR is the compiler role, this function is actually a compiler and should have its own module and be called PIL compiler.

Reply:
Agreed and fixed.